Add functions to get the process name and main module range

mitchell-merry opened the pull request:
This is something I needed to access for my scene manager stuff (need to access the range of the main module without changing the API of SceneManager::attach.

I imagine it could be useful elsewhere too.

I'm inferring the name of the main module is the same as the name of the process - I think this is a safe assumption, but let me know if not.

mitchell-merry committed:
Add functions to get the process name and main module range.

mitchell-merry committed:
Add functions to get the process name and main module range.

Jujstme commented:
Two quick observations:
1. You are allocating a string, requiring those functions to be hidden behind a `#[cfg(feature = "alloc")]`, however you can probably use an`ArrayCString` instead.
2. If I remember correctly, despite the fixes pushed by Cryze recently on the runtime, `Process::get_path()` can still fail if the process is inside special directories.

CryZe commented on src/runtime/process.rs:
```diff
@@ -161,6 +161,14 @@ impl Process {
+    #[inline]
+    pub fn get_name(&self) -> Result<alloc::string::String, Error> {
+        let path = self.get_path()?;
+        Ok(path.split("/").last().ok_or(Error {})?.into())
```
Something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=f38f77fdb69e99000a9bc52338b5738d

mitchell-merry replied:
Oh neat, let me have a play with this...

CryZe replied:
I'd say so. In both cases it needs to move the remaining string to a new location, but at least it doesn't need to allocate like this.

mitchell-merry commented:
> Two quick observations:
> 
> 1. You are allocating a string, requiring those functions to be hidden behind a `#[cfg(feature = "alloc")]`, however you can probably use an`ArrayCString` instead.
> 2. If I remember correctly, despite the fixes pushed by Cryze recently on the runtime, `Process::get_path()` can still fail if the process is inside special directories.

1. Yeah, but I'm calling get_path anyways which already allocates a string, so I don't think there's a way around it. Besides, I don't find ArrayCStrings very practical to use, I almost always want a string (if there's a way around it, I'm happy to introduce a `get_name_as_string` version instead)

2. Special directories, as in network drives? Seems like something that needs to be fixed separately still and not something I need to consider here - my first choice would have been to store the process name on the type itself so we don't need to re-calculate it, but Process is not a struct.

mitchell-merry committed:
remove an unnecessary allocation

mitchell-merry committed:
remove an unnecessary allocation

mitchell-merry commented:
Does this need anything else to go through? I think I addressed the review

I want to upstream some more of my changes before I go insane maintaining mulitple forks, so would be nice to be able to start getting things through
